fix: Better error message for dashboard import

smacker opened the pull request:
### CATEGORY

Choose one

- [x] Bug Fix
- [ ] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY

Common mistake is trying to import dashboard without creating
datasources first. Currently it causes error 500 with a message

> sqlalchemy.orm.exc.NoResultFound: No row was found for one()

which is difficult to understand.

This commit catches NoResultFound error and returns human readable error
using flash('danger').

Ref: #2992

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

<img width="1095" alt="Screenshot 2019-05-30 at 16 04 50" src="https://user-images.githubusercontent.com/406916/58638611-8fa40400-82f5-11e9-9c84-952fe3140c7d.png">
<img width="1122" alt="Screenshot 2019-05-30 at 16 03 59" src="https://user-images.githubusercontent.com/406916/58638615-929ef480-82f5-11e9-936d-6d9530474c1d.png">

### TEST PLAN

Export dashboard on one instance of Superset. Try to import it on another one without creating datasources first.

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [x] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS

smacker committed:
fix: Better error message for dashboard import

Common mistake is trying to import dashboard without creating
datasources first. Currently it causes error 500 with a message

> sqlalchemy.orm.exc.NoResultFound: No row was found for one()

which is difficult to understand.

This commit catches NoResultFound error and returns human readable error
using flash('danger').

Ref: #2992

smacker commented:
@mistercrunch @betodealmeida could you please take a look?

betodealmeida commented on superset/connectors/sqla/models.py:
```diff
@@ -945,8 +946,12 @@ def lookup_sqlatable(table):
+                return db.session.query(Database).filter_by(
+                    database_name=table.params_dict['database_name']).one()
+            except NoResultFound:
+                raise Exception(_("Database '{}' is not found".format(
```
I think it would be better to raise a custom exception here:

```python
raise DatabaseNotFound(...)
```

And then in `superset/views/core.py` we can be more specific:

```python
try:
    dashboard_import_export.import_dashboards(db.session, f.stream)
except DatabaseNotFound as e:
    flash(_(
        'Cannot import dashboard: {}\n'
        'Make sure to create the database before '
        'importing the dashboard.'
    ).format(e), 'danger')
except Exception as e:
    flash(_(
        'An unknown error occurred. '
        'Please contact your Superset administrator'
    ), 'danger')
```

My personal rule for error messages is that (1) they should be as specific as possible and (2) they should have an action that the user can take (even if it's just contacting the admin).

betodealmeida commented on superset/views/core.py:
```diff
@@ -1271,7 +1271,10 @@ def import_dashboards(self):
+            try:
+                dashboard_import_export.import_dashboards(db.session, f.stream)
+            except Exception as e:
+                flash(_('Can not import dashboard: {}').format(e), 'danger')
```
While "can not" is not wrong, "cannot" is much more usual.

betodealmeida reviewed (commented):
@smacker thanks for working on this, it's a huge improvement. I added a few comments on how to improve it a little bit more.

smacker commented:
Thank you @betodealmeida !

It is my first PR to superset I was trying to keep it as minimal as possible.
I rebased the PR on master and added new commit with your suggestions.

mistercrunch commented on superset/views/core.py:
```diff
@@ -1288,7 +1288,19 @@ def import_dashboards(self):
+                flash(_(
+                    'Cannot import dashboard: {}.\n'
+                    'Make sure to create the database before '
+                    'importing the dashboard.',
```
For `i18n` strings, the proper formatting is as documented here https://pythonhosted.org/Flask-Babel/#gettext-functions

betodealmeida replied:
My bad, I gave a bad example when reviewing this initially.

smacker replied:
>  the proper formatting

thanks! fixed.

> My bad

not at all. My original code had the same problem.

smacker commented:
Rebased to solve conflicts with recent re-format of all the code by Black.

smacker commented:
Rebased again on master to solve conflicts.
@mistercrunch @betodealmeida could you take another look, please?
